diff -Naur binutils-2.26/bfd/config.bfd binutils-2.26-patched/bfd/config.bfd
--- binutils-2.26/bfd/config.bfd	2015-11-13 08:27:40.000000000 +0000
+++ binutils-2.26-patched/bfd/config.bfd	2016-06-06 22:10:40.692744000 +0000
@@ -675,6 +675,11 @@
     targ_selvecs="i386_elf32_vec iamcu_elf32_vec l1om_elf64_vec k1om_elf64_vec"
     want64=true
     ;;
+  x86_64-*-ebbrt*)
+    targ_defvec=x86_64_elf64_vec
+    targ_selvecs=i386_elf32_vec
+    want64=true
+    ;;
   x86_64-*-elf*)
     targ_defvec=x86_64_elf64_vec
     targ_selvecs="i386_elf32_vec iamcu_elf32_vec x86_64_elf32_vec l1om_elf64_vec k1om_elf64_vec"
diff -Naur binutils-2.26/config.sub binutils-2.26-patched/config.sub
--- binutils-2.26/config.sub	2015-11-13 08:27:41.000000000 +0000
+++ binutils-2.26-patched/config.sub	2016-06-06 22:05:03.654942338 +0000
@@ -1379,7 +1379,7 @@
 	      | -hpux* | -unos* | -osf* | -luna* | -dgux* | -auroraux* | -solaris* \
 	      | -sym* | -kopensolaris* | -plan9* \
 	      | -amigaos* | -amigados* | -msdos* | -newsos* | -unicos* | -aof* \
-	      | -aos* | -aros* | -cloudabi* \
+	      | -aos* | -aros* | -cloudabi* | -ebbrt* \
 	      | -nindy* | -vxsim* | -vxworks* | -ebmon* | -hms* | -mvs* \
 	      | -clix* | -riscos* | -uniplus* | -iris* | -rtu* | -xenix* \
 	      | -hiux* | -386bsd* | -knetbsd* | -mirbsd* | -netbsd* \
diff -Naur binutils-2.26/gas/configure.tgt binutils-2.26-patched/gas/configure.tgt
--- binutils-2.26/gas/configure.tgt	2015-11-13 08:27:41.000000000 +0000
+++ binutils-2.26-patched/gas/configure.tgt	2016-06-06 22:11:53.357028254 +0000
@@ -217,6 +217,7 @@
   i386-*-beospe*)			fmt=coff em=pe ;;
   i386-*-beos*)				fmt=elf ;;
   i386-*-coff)				fmt=coff ;;
+  i386-*-ebbrt*)                         fmt=elf ;;
   i386-*-elfiamcu)			fmt=elf arch=iamcu ;;
   i386-*-elf*)				fmt=elf ;;
   i386-*-kaos*)				fmt=elf ;;
diff -Naur binutils-2.26/ld/Makefile.am binutils-2.26-patched/ld/Makefile.am
--- binutils-2.26/ld/Makefile.am	2015-11-13 08:27:42.000000000 +0000
+++ binutils-2.26-patched/ld/Makefile.am	2016-06-06 22:42:48.408798825 +0000
@@ -1987,6 +1987,10 @@
 eelf_x86_64.c: $(srcdir)/emulparams/elf_x86_64.sh \
   $(ELF_DEPS) $(srcdir)/scripttempl/elf.sc ${GEN_DEPENDS}
 
+eelf_x86_64_ebbrt.c: $(srcdir)/emulparams/elf_x86_64_ebbrt.sh \
+  $(ELF_DEPS) $(srcdir)/scripttempl/ebbrt.sc ${GEN_DEPENDS}
+	${GENSCRIPTS} elf_x86_64_ebbrt "$(tdir_elf_x86_64_ebbrt)"
+
 eelf_x86_64_cloudabi.c: $(srcdir)/emulparams/elf_x86_64_cloudabi.sh \
   $(srcdir)/emulparams/elf_x86_64.sh \
   $(ELF_DEPS) $(srcdir)/scripttempl/elf.sc ${GEN_DEPENDS}
diff -Naur binutils-2.26/ld/configure.tgt binutils-2.26-patched/ld/configure.tgt
--- binutils-2.26/ld/configure.tgt	2015-11-13 08:27:42.000000000 +0000
+++ binutils-2.26-patched/ld/configure.tgt	2016-06-06 22:19:46.514582516 +0000
@@ -256,6 +256,9 @@
 			targ64_extra_libpath=elf_x86_64
 			targ_extra_libpath=elf32_x86_64
 			tdir_i386linux=${targ_alias}aout ;;
+x86_64-*-ebbrt*)        targ_emul=elf_x86_64_ebbrt
+			targ_extra_emuls="elf_x86_64 elf_i386"
+			;;
 x86_64-*-linux-gnux32)	targ_emul=elf32_x86_64
 			targ_extra_emuls="elf_x86_64 elf_i386 elf_iamcu i386linux elf_l1om elf_k1om"
 			targ_extra_libpath="elf_i386 elf_iamcu elf_x86_64 elf_l1om elf_k1om"
diff -Naur binutils-2.26/ld/emulparams/elf_x86_64_ebbrt.sh binutils-2.26-patched/ld/emulparams/elf_x86_64_ebbrt.sh
--- binutils-2.26/ld/emulparams/elf_x86_64_ebbrt.sh	1970-01-01 00:00:00.000000000 +0000
+++ binutils-2.26-patched/ld/emulparams/elf_x86_64_ebbrt.sh	2016-06-06 22:21:40.567382376 +0000
@@ -0,0 +1,4 @@
+. ${srcdir}/emulparams/elf_x86_64.sh
+GENERATE_SHLIB_SCRIPT=no
+GENERATE_PIE_SCRIPT=no
+SCRIPT_NAME=ebbrt
diff -Naur binutils-2.26/ld/scripttempl/ebbrt.sc binutils-2.26-patched/ld/scripttempl/ebbrt.sc
--- binutils-2.26/ld/scripttempl/ebbrt.sc 00:00:00.000000000 +0000
+++ binutils-2.26-patched/ld/scripttempl/ebbrt.sc	2016-06-06 22:21:40.567382376 +0000
@@ -0,0 +1,111 @@
+# Copyright (C) 2014-2015 Free Software Foundation, Inc.
+# 
+# Copying and distribution of this file, with or without modification,
+# are permitted in any medium without royalty provided the copyright
+# notice and this notice are preserved.
+#
+cat <<EOF
+ENTRY(start32)
+
+mb_magic = 0x1badb002;
+mb_flags = 0x3;
+mb_checksum = -(mb_magic + mb_flags);
+
+SECTIONS
+{
+	. = 0x100000;
+	kbegin = .;
+	.text BLOCK(4K) : ALIGN(4K)
+	{
+		LONG(mb_magic)
+		LONG(mb_flags)
+		LONG(mb_checksum)
+		*(.text.*)
+	}
+
+	.rodata BLOCK(4K) : ALIGN(4K)
+	{
+		*(.rodata.*)
+	}
+
+        .eh_frame_hdr ALIGN(4K) : {
+                *(.eh_frame_hdr)
+        }
+
+        .eh_frame ALIGN(4) : ONLY_IF_RO {
+                  __eh_frame_start = .;
+                  *(.eh_frame)
+                  QUAD(0);
+        }
+
+        .gcc_except_table BLOCK(4K) : ALIGN(4K) ONLY_IF_RO {
+                          *(.gcc_except_table .gcc_except_table.*)
+        }
+
+        .note : ALIGN(16) {
+                *(.note*)
+        }
+
+        .ctors BLOCK(4K) : ALIGN(4K) {
+               start_ctors = .;
+               /* KEEP(*(SORT_BY_INIT_PRIORITY(.ctors.*))) */
+               KEEP(*(SORT(.ctors.*)))
+               KEEP(*(.ctors))
+               end_ctors = .;
+        }
+
+        .dtors BLOCK(4K) : ALIGN(4K) {
+               start_dtors = .;
+               KEEP(*(SORT(.dtors.*)))
+               KEEP(*(.dtors))
+               end_dtors = .;
+        }
+
+        .eh_frame ALIGN(4) : ONLY_IF_RW {
+                  __eh_frame_start = .;
+                  *(.eh_frame)
+                  QUAD(0);
+        }
+
+	.data BLOCK(4K) : ALIGN(4K)
+	{
+		*(.data.*)
+	}
+
+        . = ALIGN(64);
+        .tdata  :
+        {
+                tls_start = ABSOLUTE(.);
+                *(.tdata)
+                *(.tdata.*)
+                *(.gnu.linkonce.td.*)
+        }
+
+        .tbss :
+        {
+                *(.tbss)
+                *(.tbss.*)
+                *(.gnu.linkonce.tb.*)
+                . = ALIGN(64);
+                tls_end = ABSOLUTE(.);
+        }
+
+	.bss BLOCK(4K) : ALIGN(4K)
+	{
+                tcb0 = ABSOLUTE(.);
+                . += (tls_end - tls_start) + 8;
+		*(COMMON)
+                *(.bss)
+		*(.bss.*)
+                *(.gnu.linkonce.b*)
+	}
+
+	/DISCARD/ :
+	{
+		*(.note.*);
+	}
+
+	. = ALIGN(4K);
+	kend = .;
+}
+EOF
diff -Naur binutils-2.26/bfd/elf64-x86-64.c binutils-2.26-patched/bfd/elf64-x86-64.c
--- binutils-2.26/bfd/elf64-x86-64.c	2015-11-13 08:27:40.000000000 +0000
+++ binutils-2.26-patched/bfd/elf64-x86-64.c	2016-07-18 18:54:40.061983947 -0400
@@ -6326,7 +6326,7 @@
 #define ELF_ARCH         bfd_arch_i386
 #define ELF_TARGET_ID          X86_64_ELF_DATA
 #define ELF_MACHINE_CODE       EM_X86_64
-#define ELF_MAXPAGESIZE          0x200000
+#define ELF_MAXPAGESIZE          0x1000
 #define ELF_MINPAGESIZE          0x1000
 #define ELF_COMMONPAGESIZE       0x1000
